crud.go

```diff
@@ -152,6 +153,203 @@ func (cc *CrudComponent) GetReplica(ctx context.Context, opts *GetReplicaOptions
+			Cas:       resp.Cas,
+			IsReplica: false,
+		}, nil
+	}
+
+	getReplicaFn := func(collectionID uint32, vbID uint16, client KvClient) (*GetAllReplicasResult, error) {
+		resp, err := client.GetReplica(ctx, &memdx.GetReplicaRequest{
+			CollectionID: collectionID,
+			Key:          opts.Key,
+			VbucketID:    vbID,
+			CrudRequestMeta: memdx.CrudRequestMeta{
+				OnBehalfOf: opts.OnBehalfOf,
+			},
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		value, datatype, err := cc.compression.Decompress(memdx.DatatypeFlag(resp.Datatype), resp.Value)
+		if err != nil {
+			return nil, err
+		}
+
+		return &GetAllReplicasResult{
+			Value:     value,
+			Flags:     resp.Flags,
+			Datatype:  datatype,
+			Cas:       resp.Cas,
+			IsReplica: true,
+		}, nil
+	}
+
+	routingFn := func(i int, collectionID uint32) (*GetAllReplicasResult, error) {
+		return OrchestrateMemdRouting(ctx, cc.vbs, cc.nmvHandler, opts.Key, uint32(i), func(endpoint string, vbID uint16) (*GetAllReplicasResult, error) {
+			return OrchestrateMemdClient(ctx, cc.connManager, endpoint, func(client KvClient) (*GetAllReplicasResult, error) {
+
+				var res *GetAllReplicasResult
+				var err error
+
+				if i == 0 {
+					res, err = getFn(collectionID, vbID, client)
+				} else {
+					res, err = getReplicaFn(collectionID, vbID, client)
+				}
+
+				if res != nil && err == nil {
+					results[endpoint] = *res
+					result.OutCh <- &ReplicaStreamEntry{
+						Err: nil,
+						Res: res,
+					}
+				}
+
+				return res, err
+			})
+		})
+	}
+
+	go func() {
+		defer func() {
+			close(result.OutCh)
+		}()
+		_, err := OrchestrateMemdRetries(
+			ctx, cc.retries, func() (map[string]GetAllReplicasResult, error) {
+				return OrchestrateMemdCollectionID(
+					ctx, cc.collections, opts.ScopeName, opts.CollectionName,
+					func(collectionID uint32, manifestID uint64) (map[string]GetAllReplicasResult, error) {
+						for i := 0; i <= 3; i++ {
+							_, err := routingFn(i, collectionID)
+
+							if errors.Is(err, ErrInvalidReplica) {
+								// We have fetched num_replicas number of the docs from different servers so we can begin
+								// the end of the algo
+								if len(results) == i {
+									// Check which server holds the master copy.
+									endpoint, _, err := cc.vbs.DispatchByKey(opts.Key, 0)
+									if err != nil {
+										return nil, err
+									}
+
+									res, ok := results[endpoint]
+
+									// The master node has remained the same as when we fetched the doc from it so we are done
+									if ok && !res.IsReplica {
+										result.OutCh <- &ReplicaStreamEntry{
+											Err: nil,
+											Res: nil,
+										}
+
+										return results, nil
+									}
+
+									// Here we know that len(results) == i == num_replicas
+									// So we have fetched a sufficient number of replicas, now we just need to get the original
+									_, err = routingFn(0, collectionID)
+									if err != nil {
+										result.OutCh <- &ReplicaStreamEntry{
```

Comment:
Are we are okay returning a replica that no longer exists on that node? Say we get the original from node1 and then a replica from node2. Now a rebalance occurs and node2 becomes the master while node3 has the replica. We'd then fetch the new original from node2 and return. So the user would receive the replica that was initially on node2 and then the master from node2.

Reply:
I'm going to put some more analysis of these situations into the ING for later reference on our thought process as this appears that it's a lot more complicated than it appeared on the surface.